docs/scopes.md

```diff
@@ -0,0 +1,199 @@
+* `https://identity.mozilla.com/apps/oldsync/history#write` indicates
+   write-only access to the user's history data in Firefox Sync.
+
+To be a valid scope value, the URL must:
+
+* Be an absolute `https://` URL.
+* Have no username, password, or query component.
+* If present, have a fragment component consisting only of alphanumeric ascii characters and underscore.
+* Remain unchanged when parsed and serialized following the rules in the
+  [WhatWG URL Spec](https://url.spec.whatwg.org).
+
+The following URL scope values are currently recognized by FxA:
+
+* `https://identity.mozilla.com/apps/oldsync`: access to data in Firefox Sync.
+* `https://identity.mozilla.com/apps/notes`: access to data in Firefox Notes.
+
+
+## Scope Matching and Implication
+
+We say that a scope value A *implies* another scope value B
+if they are exactly equal,
+of if A represents a more general capability than B.
+Similarly, a scope A implies scope value B
+if there is some scope value in A that implies B.
+This is the basic operation used to check
+permissions when processing an OAuth token.
+
+Consumers of OAuth tokens should avoid
+directly parsing and comparing scopes where possible,
+and instead use the existing implementation
+in the `fxa-shared` node module.
+
+For consumers that must implement their own scope checking,
+the rules for implication can be summarized as:
+
+* For URL scope values, A implies B if A is a parent resource of B.
+* For short-name scope values, split on the ":" character,
+  and A implies B if either:
+  * B[-1] is not "write" and A is a prefix of B, or.
+  * A[-1] is "write", and:
+    * A[:-1] is a prefix of B, or
+    * B[-1] is "write" and A[:-1] is a prefix of B[:-1]
+
+More precisely, the algoritm for checking implication is:
+
+* If A is a `https://` URL, then:
+  * If B is not a `https://` URL, then fail.
+  * If the origin of B is different than that of A, then fail.
+  * If the path component list of A is not a prefix of the path
+    component list of B, then fail.
+  * If A has a fragment, then:
+    * If B does not have a fragment, then fail.
+    * If B has a fragment that differs from A, then fail.
+  * Otherwise, succeed.
+* Otherwise:
+  * If B is a `https://` URL, then fail.
+  * Split A and B into components based on `:` delimiter.
+  * If the last component of B is `write`, then:
+    * If the last component of A is not `write`, then fail.
+  * If the last component of A is `write`, remove it.
+  * If A is not a prefix of B, then fail.
+  * Otherwise, succeed.
+
+Below are some testcases against which
+scope-checking code can be validated.
+
+Valid implications:
+* `profile:write` implies `profile`.
+* `profile` implies `profile:email`.
+* `profile:write` implies `profile:email`.
+* `profile:write` implies `profile:email:write`.
+* `profile:email:write` implies `profile:email`.
+* `profile profile:email:write` implies `profile:email`.
+* `profile profile:email:write` implies `profile:display_name`.
```

Comment:
Did you mean `profile profile:display_name:write` implies `profile:display_name`?

Reply:
> Did you mean profile profile:display_name:write implies profile:display_name?

`profile` implies `profile:display_name` (both read-only), the `profile:email:write` is needed to write to the email.